polkadot/xcm/xcm-executor/src/lib.rs

```diff
@@ -1240,7 +1240,17 @@ impl<Config: config::Config> XcmExecutor<Config> {
 						// any subsequent fees
 						message.push(PayFees { asset: fees });
 					} else {
-						// unpaid execution
+						// Check if this origin is allowed to send messages that do not
+						// intend to pay for fees.
+						ensure!(
+							Config::FeeManager::is_waived(self.origin_ref(), FeeReason::InitiateTransfer),
+							XcmError::BadOrigin
+						);
+
+						// We push the UnpaidExecution instruction to notify we do not intend to pay
+						// for fees.
+						// The receiving chain must decide based on the origin of the message if they
+						// accept this.
 						message
 							.push(UnpaidExecution { weight_limit: Unlimited, check_origin: None });
```

Comment:
This is not correct afaict: `Config::FeeManager::is_waived` will specify origins whose fee is waived on _the local_ chain; whereas the instructions pushed to `message` here are meant to be executed on _the remote_ chain.

For local fees, `Config::FeeManager::is_waived` is already being checked by the inner implementation of `fn take_fees()`.

What this PR should be changing the building of remote `message` to move pushing `UnpaidExecution` later in the remote program:
- if `preserve_origin` append this `UnpaidExecution` _just after_ `AliasOrigin` 
- if not append this `UnpaidExecution` _just before_ `ClearOrigin`